src/charts/charts.utils.ts

```
@@ -1,13 +1,122 @@
+    case ChartTimeFrame.ThreeMonth:
+    case ChartTimeFrame.Month:
+    case ChartTimeFrame.YTD:
+      update = difference >= ONE_DAY_MS;
+      break;
+    case ChartTimeFrame.Week:
+      update = difference >= ONE_HOUR_MS * 6;
+      break;
+    default:
+      update = difference >= ONE_HOUR_MS;
+  }
+  return update;
+}
+
+export function shouldTrim(reference: number, timestamp: number, timeframe: ChartTimeFrame): boolean {
+  const difference = reference - timestamp;
+  let update;
+  switch (timeframe) {
+    case ChartTimeFrame.Year:
+      update = difference >= ONE_DAY_MS * 365;
+      break;
+    case ChartTimeFrame.ThreeMonth:
+      update = difference >= ONE_DAY_MS * 90;
+      break;
+    case ChartTimeFrame.Month:
+      update = difference >= ONE_DAY_MS * 30;
+      break;
+    case ChartTimeFrame.Week:
+      update = difference >= ONE_DAY_MS * 7;
+      break;
+    case ChartTimeFrame.Day:
+      update = difference >= ONE_DAY_MS;
+      break;
+    // Year to date needs a full clear - so we will never trim
+    case ChartTimeFrame.YTD:
+    default:
+      update = false;
+  }
+  return update;
+}
+
+export async function updateSnapshots<T extends ChartData>(namespace: string, snapshot: T) {
+  const mapper = getDataMapper();
+
+  const { id, timestamp: now } = snapshot;
+  for (const timeframe of Object.values(ChartTimeFrame)) {
+    const searchKey = Object.assign(new ChartDataBlob<T>(), {
+      id: toChartDataKey(namespace, id, timeframe),
+    });
+
+    let cachedChart: ChartDataBlob<T> | undefined;
+    try {
+      cachedChart = await mapper.get(searchKey);
+      if (timeframe === ChartTimeFrame.YTD) {
+        const date = new Date(now);
+        const isFirstOfYear = date.getDay() === 0 && date.getMonth() === 0;
+        if (isFirstOfYear && cachedChart.data.length > 1) {
+          // new year, force a new object to be created
+          cachedChart = undefined;
+        }
+      }
+    } catch {} // no item found
```

May want to log something here.  Could be missing some errors with cache if it fails silently